### PR TITLE
Don't rely on router.getOperaionCount to be zero in tests

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
In tests, let's compare the operation counts before and after tests.

### CODE REVIEW

File: ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java [hunk 1]
```diff
@@ -356,8 +358,7 @@ public void testBlobNotFoundReturnedWhenAllReplicasReturnNotFound() throws Excep
     RouterErrorCode expectedErrorCode = RouterErrorCode.BlobDoesNotExist;
     FutureResult<Void> future = new FutureResult<>();
     TestCallback<Void> callback = new TestCallback<>();
-    deleteManager.submitDeleteBlobOperation(blobIdString, LOCAL_DC, future, callback,
-        quotaChargeCallback);
+    deleteManager.submitDeleteBlobOperation(blobIdString, LOCAL_DC, future, callback, quotaChargeCallback);
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java:361]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java [hunk 2]
```diff
@@ -132,8 +132,7 @@ static void testWithErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, Partit
    * @param errorCodeChecker Performs the checks that ensure that {@link RouterErrorCode#TooManyRequests} is returned .
    * @throws Exception
    */
-  static void testWithQuotaRejection(ErrorCodeChecker errorCodeChecker)
-      throws Exception {
+  static void testWithQuotaRejection(ErrorCodeChecker errorCodeChecker) throws Exception {
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java:135]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java [hunk 3]
```diff
@@ -312,7 +312,7 @@ static void compareContent(byte[] originalContent, ByteRange range, ReadableStre
   static List<ChunkInfo> buildChunkList(ClusterMap clusterMap, BlobId.BlobDataType blobDataType, long ttl,
       LongStream chunkSizeStream) {
     return chunkSizeStream.mapToObj(
-        chunkSize -> new ChunkInfo(getRandomBlobId(clusterMap, blobDataType), chunkSize, ttl))
+            chunkSize -> new ChunkInfo(getRandomBlobId(clusterMap, blobDataType), chunkSize, ttl))
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java:315]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java [hunk 4]
```diff
@@ -373,14 +374,12 @@ private void assertDeleted(String blobId) throws Exception {
   }
 
   private void assertNotDeleted(String blobId) throws Exception {
-    router.getBlob(blobId, new GetBlobOptionsBuilder().build())
-        .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    router.getBlob(blobId, new GetBlobOptionsBuilder().build()).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java:377]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java [hunk 5]
```diff
@@ -373,14 +374,12 @@ private void assertDeleted(String blobId) throws Exception {
...
   }
 
   private void deleteBlobInAllServer(String blobId) throws Exception {
     // Delete this blob
-    router.deleteBlob(blobId, "serviceid")
-        .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+    router.deleteBlob(blobId, "serviceid").get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java:382]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java [hunk 6]
```diff
@@ -363,8 +365,7 @@ private Properties getNonBlockingRouterProperties(int parallelism) {
 
   private void assertDeleted(String blobId) throws Exception {
     try {
-      router.getBlob(blobId, new GetBlobOptionsBuilder().build())
-          .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      router.getBlob(blobId, new GetBlobOptionsBuilder().build()).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java:368]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java [hunk 7]
```diff
@@ -151,8 +154,7 @@ public void basicThroughRouterTest() throws Exception {
     for (String blobId : blobIds) {
       deleteBlobInAllServer(blobId);
       // Undelete requires global quorum, so we have to make sure all the mock servers received a delete.
-      router.undeleteBlob(blobId, UNDELETE_SERVICE_ID)
-          .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      router.undeleteBlob(blobId, UNDELETE_SERVICE_ID).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java:157]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java [hunk 8]
```diff
@@ -468,8 +467,9 @@ private void sendRequestsGetResponse(FutureResult<Void> future, UndeleteManager
 
       List<ResponseInfo> responseInfoList;
       if (isQuotaRejected) {
-        responseInfoList = requestInfoList.stream().map(requestInfo -> new ResponseInfo(requestInfo, true)).collect(
-            Collectors.toList());
+        responseInfoList = requestInfoList.stream()
+            .map(requestInfo -> new ResponseInfo(requestInfo, true))
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/UndeleteManagerTest.java:471]: rm formatting edits. 
PR can be smaller if all formatting edits are moved to a different PR.

File: ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java [hunk 9]
```diff
@@ -198,13 +197,14 @@ static void setServerErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, Parti
...
-  static void assertCloseCleanup(NonBlockingRouter router) {
+  static void assertCloseCleanup(NonBlockingRouter router, long operationCount) {
     router.close();
     Assert.assertEquals("No ChunkFiller Thread should be running after the router is closed", 0,
         TestUtils.numThreadsByThisName("ChunkFillerThread"));
     Assert.assertEquals("No RequestResponseHandler should be running after the router is closed", 0,
         TestUtils.numThreadsByThisName("RequestResponseHandlerThread"));
-    Assert.assertEquals("All operations should have completed", 0, router.getOperationsCount());
+    long operationCountAfterTest = router.getOperationsCount();
+    Assert.assertEquals("All operations should have completed", operationCount, operationCountAfterTest);
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java:207]: if all operations should have completed, why isn't the op count 0 ?

**Comment by snalli** [comment]
A lot of formatting edits in here that take the attention away from core changes.

File: ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java [hunk 10]
```diff
@@ -130,20 +131,21 @@ public void init() throws Exception {
...
     deleteManager =
-        new DeleteManager(clusterMap, new ResponseHandler(clusterMap), accountService, notificationSystem,
-            routerConfig, routerMetrics, new RouterCallback(null, new ArrayList<>()), mockTime);
+        new DeleteManager(clusterMap, new ResponseHandler(clusterMap), accountService, notificationSystem, routerConfig,
+            routerMetrics, new RouterCallback(null, new ArrayList<>()), mockTime);
     MockNetworkClientFactory networkClientFactory =
         new MockNetworkClientFactory(vProps, mockSelectorState, MAX_PORTS_PLAIN_TEXT, MAX_PORTS_SSL,
             CHECKOUT_TIMEOUT_MS, serverLayout, mockTime);
     networkClient = networkClientFactory.getNetworkClient();
+    operationCountBeforeTest = router.getOperationsCount();
```
Comment (snalli) [inline ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java:140]: how can we know op count before test ? Also shouldn't it be deterministically zero since we just started the test ?

File: ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java [hunk 11]
```diff
@@ -198,13 +197,14 @@ static void setServerErrorCodes(ServerErrorCode[] serverErrorCodesInOrder, Parti
   /**
    * Asserts that expected threads are not running after the router is closed.
    */
-  static void assertCloseCleanup(NonBlockingRouter router) {
```
Comment (JingQianCloud) [inline ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java:201]: I may prefer we keep the original assertCloseCleanup function without the operationCount since zero is the ideal case.
We just add the new one with the operationCount.

File: ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java (hunk 10 above)
Comment (justinlin-linkedin) [reply ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java:140]: the issue with operation count is that it's a static and yet mutable variable. So router.getOperationsCount() would just return this value. Since it's static, different test cases would interfere each other. If there is one test that increases operation count but didn't wait for it to finish, then operation count would not be zero any more for other tests. With this change, we would compare operation counts before and after tests and to make sure the tests that care about the operation counts, have the proper way to verify all the operations are done.

File: ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java (hunk 1 above)
Comment (justinlin-linkedin) [reply ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java:361]: I don't mind these formatting change, it's pretty small PR anyway.

**Review by JingQianCloud — APPROVED** [review]
Thanks for fixing the TtlUpdateManageTest failure issue.

The code isolated the different unit tests and doesn't allow the errors from the previous unit test impacts the following ones.

File: ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java (hunk 11 above)
Comment (justinlin-linkedin) [reply ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java:201]: I have different opinion. We should not by default assume the operation count before test is 0 if we really care about the isolation of different unit tests. We want to make sure there is no unfinished operation created by the current unit test, then the correct way of doing it, is to compare the before and after operation counts.
Even if we fixed all the unit tests that break the operation count, we still would prefer having a before operation count to compare with when we clean up for the current unit test.

File: ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java (hunk 1 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java:361]: could you please rm them ? its just clutter

**Review by snalli — CHANGES_REQUESTED** [review]
please rm formatting changes.

File: ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java (hunk 9 above)
Comment (mikeyang923) [reply ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java:207]: Looks it will be non-zero if the other test is in the middle (not finished yet).

File: ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java (hunk 10 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/DeleteManagerTest.java:140]: I see, please put this in the commit description so when look at the history we know the reason for this change.

File: ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java (hunk 9 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/RouterTestHelpers.java:207]: We run tests serially as I recall. So we can't be in the middle of some test while running another.

**Review by SophieGuo410 — APPROVED** [review]
